### PR TITLE
Symbolic

### PR DESCRIPTION
**Pull request by pehamTom**
## Description

This PR introduced functionality for equivalence checking of parametrized circuits with QCEC. This PR is related to https://github.com/cda-tum/qfr/pull/199 and https://github.com/cda-tum/zx/pull/15 

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: mqt/qcec/symbolic.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        return (qc1_bound, qc2_bound)
+
+    for i in range(0, n_checks - 1):
+
+        circ1_inst, circ2_inst = instantiate_params(i, circ1, circ2, **kwargs)
+        res = check_instantiated(circ1_inst, circ2_inst)
+        if res.equivalence == "not_equivalent":
+            return res
+
+    res = check_instantiated_random(circ1, circ2, parameters, **kwargs)
```
Comment (burgholzer) [inline mqt/qcec/symbolic.py:145]: following up on the comment above, this could probably be a bit more transparent.
there is _one_ run with precisely determined RHS and _one_ last resort random instantiation.
in-between, there are the other attempts.

File: .gitmodules [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [submodule "extern/qfr"]
-	path = extern/qfr
-	url = https://github.com/cda-tum/qfr.git
-    branch = main
+        path = extern/qfr
+        url = https://github.com/cda-tum/qfr.git
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply .gitmodules:3]: Not intentional. Fixed.

File: src/EquivalenceCheckingManager.cpp [hunk 3]
```diff
@@ -6,10 +6,13 @@
 #include "EquivalenceCheckingManager.hpp"
 
 #include "EquivalenceCriterion.hpp"
+#include "Expression.hpp"
 #include "zx/FunctionalityConstruction.hpp"
 
 #include <iostream>
 #include <memory>
+#include <ostream>
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply src/EquivalenceCheckingManager.cpp:14]: True. Fixed.

File: src/EquivalenceCheckingManager.cpp [hunk 4]
```diff
@@ -174,9 +182,10 @@ EquivalenceCheckingManager::EquivalenceCheckingManager(
   // set numeric tolerance used throughout the check
   setTolerance(configuration.execution.numericalTolerance);
 
-  // run all configured optimization passes
-  runOptimizationPasses();
-
+  if (qc1.isVariableFree() && qc2.isVariableFree()) {
+    // run all configured optimization passes
+    runOptimizationPasses();
+  }
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply src/EquivalenceCheckingManager.cpp:188]: It is indeed a bit tricky. I would suggest making it an issue and fixing it after the 2.0 release. What do you say?

File: src/EquivalenceCheckingManager.cpp [hunk 5]
```diff
@@ -191,16 +200,17 @@ EquivalenceCheckingManager::EquivalenceCheckingManager(
                  "inputs! Proceed with caution!\n";
   }
 
-  // try to fix a potential mismatch in the output permutations of both circuits
+  // try to fix a potential mismatch in the output permutations of both
+  // circuits
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply src/EquivalenceCheckingManager.cpp:204]: Fixed it anyway. That is indeed from a config in emacs' lsp mode for C++.

File: test/test_symbolic.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  std::unique_ptr<QuantumComputation> symQc1 =
+      std::make_unique<QuantumComputation>(1);
+  std::unique_ptr<QuantumComputation> symQc2 =
+      std::make_unique<QuantumComputation>(1);
+
+protected:
+  // void TearDown override{};
+
+  void SetUp() override{};
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply test/test_symbolic.cpp:29]: Done

File: test/test_symbolic.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+public:
+  Variable x = Variable("x");
+
+  Symbolic xMonom    = Symbolic{Term<dd::fp>{x}};
+  Symbolic xMonomNeg = Symbolic{-Term<dd::fp>{x}};
+
+  std::unique_ptr<QuantumComputation> symQc1 =
+      std::make_unique<QuantumComputation>(1);
+  std::unique_ptr<QuantumComputation> symQc2 =
+      std::make_unique<QuantumComputation>(1);
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply test/test_symbolic.cpp:24]: Changed it.

File: test/test_symbolic.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+TEST_F(SymbolicTest, SymbolicEqu) {
+  symQc1->rx(0, xMonom);
+
+  symQc2->h(0);
+  symQc2->rz(0, xMonom);
+  symQc2->h(0);
+
+  ec::Configuration config{};
+  const auto        ecm = std::make_unique<ec::EquivalenceCheckingManager>(
+      *symQc1, *symQc2, config);
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply test/test_symbolic.cpp:41]: Changed that as well.

File: src/EquivalenceCheckingManager.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (burgholzer) [reply src/EquivalenceCheckingManager.cpp:188]: Agreed. That sounds like a good plan!

File: mqt/qcec/verify.py [hunk 9]
```diff
@@ -24,9 +32,13 @@ def verify(
     :param circ1: The first circuit.
     :param circ2: The second circuit.
     :param configuration: The configuration to use for the equivalence checking process.
+    :param n_symbolic_checks: The number of times a parametrized circuit should be instantiated and checked
+    :param symbolic_tol: The tolerance to which instantiated parameters should be recognized as 0
     :param kwargs: Keyword arguments to pass to the :class:`EquivalenceCheckingManager <.EquivalenceCheckingManager>` constructor.
     :return: The results of the equivalence checking process.
     """
+    if (not isinstance(circ1, str) and circ1.parameters) or (not isinstance(circ2, str) and circ2.parameters):
+        return check_symbolic(circ1, circ2, n_symbolic_checks, symbolic_tol, **kwargs)
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/verify.py:41]: Done.

File: mqt/qcec/symbolic.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
+"""Functionality for checking equivalence of parameterized quantum circuits."""
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:1]: As discussed in person, since parameters don't mean quite the same for quiskit and mqt. I'm going to at least streamline the parts that are exposed to the user (like `n_parameterized_checks` vs `n_symbolic_checks`)

File: mqt/qcec/symbolic.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    ecm = EquivalenceCheckingManager(circ1, circ2, **kwargs)
+    ecm.set_zx_checker(True)
+    ecm.set_alternating_checker(False)
+    ecm.set_simulation_checker(False)
+    ecm.set_parallel(False)
+
+    ecm.run()
+
+    print(circ1.draw())
+    print(circ2.draw())
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:27]: They definitely shouldn't. A debug print that snuck into the PR

File: mqt/qcec/symbolic.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+import numpy as np
+from mqt.qcec import EquivalenceCheckingManager
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter, ParameterExpression
+
+
+def check_symbolic_zx(
+    circ1: QuantumCircuit, circ2: QuantumCircuit, **kwargs: Any
+) -> EquivalenceCheckingManager.Results:
+    """Check circuits for equivalence with the zx_calculus."""
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:17]: Done

File: mqt/qcec/symbolic.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+    ecm.run()
+
+    print(circ1.draw())
+    print(circ2.draw())
+    return ecm.get_results()
+
+
+def extract_params(circ1: QuantumCircuit, circ2: QuantumCircuit) -> set[Parameter]:
+    """Extract parameters and equations of symbolic parameterized circuits."""
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:32]: Yes. Removed the "symbolic" part

File: mqt/qcec/symbolic.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    p1 = set(circ1.parameters)
+    p2 = set(circ2.parameters)
+
+    p = p1.union(p2)
+
+    n_params = len(p)
+    exprs = list(chain(*[instr[0].params for instr in circ1.data + circ2.data if instr[0].params != []]))
+
+    def is_expr(x):
+        return type(x) is Parameter or type(x) is ParameterExpression
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:42]: Yes, adjusted it to be in line with the way it is handled in QMAP

File: mqt/qcec/symbolic.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+                val = float((expr + offsets[row]).bind(one_map))
+                equs[row, col] = val
+
+    return symb_params, equs, offsets
+
+
+def check_instantiated(
+    circ1: QuantumCircuit, circ2: QuantumCircuit, **kwargs: Any
+) -> EquivalenceCheckingManager.Results:
+    """Check circuits for equivalence with dd equivalence checker."""
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:69]: Done

File: mqt/qcec/symbolic.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    ecm = EquivalenceCheckingManager(circ1, circ2, **kwargs)
+    ecm.set_zx_checker(False)
+    ecm.set_alternating_checker(True)
+    ecm.set_simulation_checker(True)
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:73]: Just used the default constructor of the EquivalenceCheckingManager and disabled the ZX checker

File: mqt/qcec/symbolic.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+    ecm.run()
+
+    print(circ1.draw())
+    print(circ2.draw())
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:78]: Again, a debug print that made its way into the PR

File: mqt/qcec/symbolic.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    circ1_inst = circ1.bind_parameters(param_map)
+    circ2_inst = circ2.bind_parameters(param_map)
+    ecm = EquivalenceCheckingManager(circ1_inst, circ2_inst, **kwargs)
+    ecm.set_zx_checker(False)
+    ecm.set_alternating_checker(True)
+    ecm.set_simulation_checker(True)
+
+    ecm.run()
+
+    return ecm.get_results()
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:99]: True. Changed it accordingly

File: mqt/qcec/symbolic.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+) -> EquivalenceCheckingManager.EquivalenceCheckingManager.Results:
+    """Equivalence checking flow for parameterized circuit."""
+    res = check_symbolic_zx(circ1, circ2, **kwargs)
+
+    if res.considered_equivalent():
+        return res
+
+    parameters, A, offsets = extract_params(circ1, circ2)
+
+    def instantiate_params(i, qc1: QuantumCircuit, qc2: QuantumCircuit) -> tuple[QuantumCircuit, QuantumCircuit]:
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:113]: Fixed

File: mqt/qcec/symbolic.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+        phases = [0, np.pi, np.pi / 2, -np.pi / 2, np.pi / 4, -np.pi / 4]
+        if i > 0:
+            b = np.random.choice(phases, size=len(offsets)) + offsets
+        else:
+            b = offsets
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/symbolic.py:118]: Good point. Changed it

File: mqt/qcec/symbolic.py (hunk 1 above)
Comment (pehamTom) [reply mqt/qcec/symbolic.py:145]: Good point. For symmetries' sake I also made a dedicated method for the first instance

File: mqt/qcec/verify.py [hunk 21]
```diff
@@ -5,12 +5,20 @@
...
+from .symbolic import check_symbolic
+
 
 def verify(
-    circ1: QuantumCircuit | str, circ2: QuantumCircuit | str, configuration: Configuration | None = None, **kwargs: Any
+    circ1: QuantumCircuit | str,
+    circ2: QuantumCircuit | str,
+    configuration: Configuration | None = None,
+    n_symbolic_checks: int = 6,
+    symbolic_tol: float = 1e-12,
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply mqt/qcec/verify.py:16]: Done. That was quite some code duplication with the bindings. As discussed in person, the configuration options are only used for the json printing on the C++ side.

File: src/EquivalenceCheckingManager.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (burgholzer) [reply src/EquivalenceCheckingManager.cpp:187]: Tracked in #168.

File: docs/source/CompilationFlowVerification.ipynb [hunk 22]
```diff
@@ -2,6 +2,7 @@
  "cells": [
   {
    "cell_type": "markdown",
+   "metadata": {},
```
Comment (burgholzer) [inline docs/source/CompilationFlowVerification.ipynb:5]: The changes throughout this file seem to arise from the execution of the Jupiter script. That does not seem intentional? could you please revert the changes in this file?

File: docs/source/index.rst [hunk 23]
```diff
@@ -25,6 +25,7 @@ We appreciate any feedback and contributions to the project. If you want to cont
     Quickstart
     EquivalenceChecking
     CompilationFlowVerification
+    ParameterizedCircuits
```
Comment (burgholzer) [inline docs/source/index.rst:28]: Is it possible that you missed uploading that file? I can't find it and it isn't shown in the docs.

File: mqt/qcec/bindings.cpp [hunk 24]
```diff
@@ -96,11 +96,15 @@ static std::unique_ptr<EquivalenceCheckingManager> createManagerFromOptions(
...
-    const std::size_t maxSims   = std::max(16U,
-                                           std::thread::hardware_concurrency() -
-                                               2U),
-    const StateType&  stateType = StateType::ComputationalBasis,
+    const std::size_t maxSims = std::max(16U,
+                                         std::thread::hardware_concurrency() -
+                                             2U),
+    // Parameterized
+    const double      parameterizedTol          = 1e-12,
+    const std::size_t nAdditionalInstantiations = 0,
```
Comment (burgholzer) [inline mqt/qcec/bindings.cpp:104]: Are these options purposely sandwiched between the "Simulation" options?

File: mqt/qcec/bindings.cpp [hunk 25]
```diff
@@ -96,11 +96,15 @@ static std::unique_ptr<EquivalenceCheckingManager> createManagerFromOptions(
...
+    const StateType&  stateType                 = StateType::ComputationalBasis,
     const std::size_t seed = 0U, const bool storeCEXinput = false,
+
```
Comment (burgholzer) [inline mqt/qcec/bindings.cpp:107]: I guess this newline is not intentional 🙃

File: mqt/qcec/bindings.cpp [hunk 26]
```diff
@@ -283,6 +291,8 @@ PYBIND11_MODULE(pyqcec, m) {
          "alternating_scheme"_a = "proportional", "profile"_a = "",
          "trace_threshold"_a = 1e-8, "fidelity_threshold"_a = 1e-8,
          "max_sims"_a = std::max(16U, std::thread::hardware_concurrency() - 2U),
+         "parameterized_tolerance"_a   = 1e-12,
+         "additional_instantiations"_a = 0U,
```
Comment (burgholzer) [inline mqt/qcec/bindings.cpp:295]: Just as a reminder that this also might need to be touched. Same question as above.

File: mqt/qcec/bindings.cpp [hunk 27]
```diff
@@ -536,6 +553,7 @@ PYBIND11_MODULE(pyqcec, m) {
       .def_readwrite("application", &Configuration::application)
       .def_readwrite("functionality", &Configuration::functionality)
       .def_readwrite("simulation", &Configuration::simulation)
+      .def_readwrite("parameterized", &Configuration::parameterized)
```
Comment (burgholzer) [inline mqt/qcec/bindings.cpp:556]: I can't link to the proper place in this review since there have been no changes there, but these options are not reflected in the docs.
A subpage needs to be added here https://github.com/cda-tum/qcec/blob/e9aeae06ee1c1cd30c759109df15c18ec2eaf476/docs/source/library/Configuration.rst
And the subpage should probably look similar to any in https://github.com/cda-tum/qcec/tree/symbolic/docs/source/library/configuration

File: mqt/qcec/parameterized.py [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+) -> EquivalenceCheckingManager:
+    if kwargs:
+        # create the equivalence checker from keyword arguments
+        return EquivalenceCheckingManager(circ1, circ2, **kwargs)
+    else:
+        if configuration is None:
+            configuration = Configuration()
+
+        # create the equivalence checker from configuration
+        return EquivalenceCheckingManager(circ1, circ2, configuration)
```
Comment (burgholzer) [inline mqt/qcec/parameterized.py:25]: At some point (not this PR), this should probably be unified in some place so that the other `verify` functions can make use of it too. 
Technically, that function could be exposed as part of the bindings. (a class method of the ECM `from_config_or_kwargs`). Now that I think of it: wouldn't that also be a better way for the, currently overloaded, regular constructors? Overloading in Python isn't that great, right?

File: mqt/qcec/parameterized.py [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+        phases = [0, np.pi, np.pi / 2, -np.pi / 2, np.pi / 4, -np.pi / 4]
+        b = np.random.choice(phases, size=len(offsets)) + offsets
+        return instantiate_params(qc1, qc2, b)
+
+    circ1_inst, circ2_inst = instantiate_params_zero(circ1, circ2)
+    res = check_instantiated(circ1_inst, circ2_inst, configuration, **kwargs)
+    if res.equivalence == "not_equivalent":
+        return res
+
+    for _i in range(1, n_checks - 1):
```
Comment (burgholzer) [inline mqt/qcec/parameterized.py:166]: shouldn't `n_checks` happen here? I thought that parameter configures the number of additional instantiations. (maybe a better name should be used here). Then, this should be `range(n_checks)`

File: mqt/qcec/verify.py [hunk 30]
```diff
@@ -28,6 +34,12 @@ def verify(
     :return: The results of the equivalence checking process.
     """
 
+    def is_parameterized(qc: QuantumCircuit) -> bool:
+        return not isinstance(qc, str) and qc.parameters
+
+    if is_parameterized(circ1) or is_parameterized(circ2):
+        return check_parameterized(circ1, circ2, configuration, **kwargs)
+
```
Comment (burgholzer) [inline mqt/qcec/verify.py:42]: Just a reminder: the `verify_compilation` function should also be enriched with these new options.
(At which point it would probably be great again to have the `is_parameterized` function available to both)

**Review by burgholzer — CHANGES_REQUESTED** [review]
I just quickly went through all the changes and added my remarks down below.

File: docs/source/index.rst (hunk 23 above)
Comment (pehamTom) [reply docs/source/index.rst:28]: Indeed. Changed it back.

File: mqt/qcec/bindings.cpp (hunk 24 above)
Comment (pehamTom) [reply mqt/qcec/bindings.cpp:104]: Put it afterwards now

File: mqt/qcec/bindings.cpp (hunk 25 above)
Comment (pehamTom) [reply mqt/qcec/bindings.cpp:107]: Don't think so. Changed it

File: mqt/qcec/bindings.cpp (hunk 26 above)
Comment (pehamTom) [reply mqt/qcec/bindings.cpp:295]: Done

File: mqt/qcec/parameterized.py (hunk 29 above)
Comment (pehamTom) [reply mqt/qcec/parameterized.py:166]: That is a leftover from how it was handled previously. Changed it back.

File: mqt/qcec/bindings.cpp (hunk 27 above)
Comment (pehamTom) [reply mqt/qcec/bindings.cpp:556]: Done

File: mqt/qcec/parameterized.py (hunk 28 above)
Comment (pehamTom) [reply mqt/qcec/parameterized.py:25]: I think the current method is actually adequate. We can simply have a `utils.py` that all files can use

File: mqt/qcec/verify.py (hunk 30 above)
Comment (pehamTom) [reply mqt/qcec/verify.py:42]: Extracted it into a function in the `parameterized` module so both verify and verify_compilation can use it

File: docs/source/CompilationFlowVerification.ipynb (hunk 22 above)
Comment (pehamTom) [reply docs/source/CompilationFlowVerification.ipynb:5]: Did that with `nb-clean` nice tool. Might be worth using in a pre-commit hook

Comment (burgholzer) [reply docs/source/CompilationFlowVerification.ipynb:5]: Is there a pre-commit check for that?

File: mqt/qcec/parameterized.py (hunk 28 above)
Comment (pehamTom) [reply mqt/qcec/parameterized.py:39]: Actually, this will be resolved by fixing Issue  #175

File: docs/source/CompilationFlowVerification.ipynb [hunk 31]
```diff
@@ -143,34 +100,26 @@
...
-    "version": 3
+    "version": 2
    },
    "file_extension": ".py",
    "mimetype": "text/x-python",
    "name": "python",
    "nbconvert_exporter": "python",
-   "pygments_lexer": "ipython3",
-   "version": "3.9.15"
+   "pygments_lexer": "ipython2"
```
Comment (burgholzer) [inline docs/source/CompilationFlowVerification.ipynb:120]: I don't know why, but this change looks strange to me. Can it be that some Python2 slipped in there somehow?

File: mqt/qcec/parameterized.py [hunk 32]
```diff
@@ -121,18 +135,55 @@ def check_parameterized(
...
+    def update_stats(res: EquivalenceCheckingManager.Results) -> None:
+        nonlocal total_preprocessing_time
+        nonlocal total_runtime
+        nonlocal total_simulations_started
+        nonlocal total_simulations_finished
+        total_preprocessing_time += res.preprocessing_time
+        total_runtime += res.check_time
+        total_simulations_started += res.started_simulations
+        total_simulations_finished += res.performed_simulations
+        return
```
Comment (burgholzer) [inline mqt/qcec/parameterized.py:152]: that statement is superfluous

**Review by burgholzer — APPROVED** [review]
Still found some really minor things while going over the code. I'll probably incorporate those over the weekend.

File: docs/source/CompilationFlowVerification.ipynb (hunk 31 above)
Comment (pehamTom) [reply docs/source/CompilationFlowVerification.ipynb:120]: Shouldn't be the case. Only let the notebook cleaner run over it

File: test/python/test_symbolic.py [hunk 33]
```diff
@@ -145,3 +162,53 @@ def test_cnot_rx_non_equ(cnot_rx: QuantumCircuit, cnot_rx_flipped: QuantumCircui
...
+    result = qcec.verify_compilation(
+        original_circuit, compiled_circuit, optimization_level, AncillaMode.NO_ANCILLA, config
+    )
+    assert (
+        result.equivalence == qcec.EquivalenceCriterion.equivalent
+        or result.equivalence == qcec.EquivalenceCriterion.equivalent_up_to_global_phase
+    )
+
+
+def test_performed_insantiations(rz_commute_lhs: QuantumCircuit, rz_commute_rhs_incorrect: QuantumCircuit) -> None:
```
[the remark this answers is not in the document]
Comment (pehamTom) [reply test/python/test_symbolic.py:203]: Done

File: mqt/qcec/parameterized.py (hunk 32 above)
Comment (pehamTom) [reply mqt/qcec/parameterized.py:152]: done

File: docs/source/CompilationFlowVerification.ipynb (hunk 31 above)
Comment (pehamTom) [reply docs/source/CompilationFlowVerification.ipynb:120]: Really don't know how the hell that happened, needed to manually change it back...
